src/native/libs/System.Security.Cryptography.Native/pal_ssl.c

```diff
@@ -577,7 +577,7 @@ X509* CryptoNative_SslGetPeerCertificate(SSL* ssl)
     long len = SSL_get_tlsext_status_ocsp_resp(ssl, &data);
     X509* cert = SSL_get1_peer_certificate(ssl);
 
-    if (len > 0 && cert != NULL)
+    if (len > 0 && cert != NULL && !X509_get_ex_data(cert, g_x509_ocsp_index))
```

Comment:
Is it possible data from `SSL_get_tlsext_status_ocsp_resp` is a "newer" staple? Should we instead free the existing data and replace it?

Reply:
Theoreticaly, the staple can be newer, but I am more afraid of introducing a possible race between another thread which would read the staple and this thread freeing it.

Reply:
Just to elaborate more, the staple may be newer only if Renegotiation happened (and server sent different staple than initially).

Reply:
Given that the original staple is very likely still valid, and renegotiation is rather rare (or so I think it is), it should be fine.

Reply:
The same concerns that would apply to the OCSP staple racing seem like they would apply to the certificate racing (the delta between when we've called SSL_get_certificate and called X509_up_ref).  And, since the certificate could have changed entirely during the renegotiate, it seems like the new OCSP staple should be preferred.

Though, since this is itself on accessing the certificate, I would find it very unlikely that OpenSSL would give back the same `X509*` after a renegotiate with different OCSP... so this is probably fine.  A new cert in renegotiation will not have pre-loaded OCSP data ready, and an old cert being requested again seems unlikely to have updated stapling.

Reply:
The difference I see there is that access to SSL is guarded by lock, so there is no race on SSL_get*_peer_certificate, and once other parts of .NET get hold of the cert, they have an up-reffed SafeHandle.

With the OCSP response, the code assumes that the OCSP_RESPONSE pointer stored in X509 is valid as long as the X509 itself is valid. Updating the stored pointer would break that assumption.

I believe the code would not crash even if we updated the OCSP_RESPONSE, since the two accesses I saw during the repro were both done from the same "thread", but better safe than sorry.